### PR TITLE
Skip analysis for unchanged files when the analysis allows it

### LINKED ISSUE #3345: Skip analysis for unchanged files when the analysis allows it
When the sensor context indicates that unchanged files can be skipped from analysis (in practice this is during PR analysis), the SonarJS analyzers should execute only the UCFG rule (from sonar-security frontend) for files marked as unchanged.

UCFG rule generates UCFG files required for security tainted analysis. The eslint-bridge loads this external rule dynamically according to the 'bundles' node command parameter and then the ESLint linter is created based on the request rules parameter.
Thus a second 'optimized' linter should be created in the initialization service with only this external rule (`ucfg`). And the JS/TS/YAML analysis services need a new flag triggering the use of the 'optimized' linter.

On the integration test side we need to validate that UCFG are properly generated for unchanged files. For that we need to add sonar-security plugin to the orchestrator and assert presence of either UCFG files (low level check) or security issues (then some security rule should be activated). Probably we should go for a first option as more reliable and controllable by us.

### PR DESCRIPTION
**Pull request by francoismora**
Fixes #3345

### CODE REVIEW

File: eslint-bridge/src/services/analysis/analyzers/js/analyzer.ts [hunk 1]
```diff
@@ -51,10 +51,12 @@ import { JsTsAnalysisInput, JsTsAnalysisOutput } from './analysis';
  * @returns the JavaScript / TypeScript analysis output
  */
 export function analyzeJSTS(input: JsTsAnalysisInput, language: Language): JsTsAnalysisOutput {
-  assertLinterInitialized();
+  debug(`analyzing file linterId=${input.linterId}`);
```
Comment (vilchik-elena) [inline eslint-bridge/src/services/analysis/analyzers/js/analyzer.ts:54]: let's add path to the file as well

File: eslint-bridge/src/services/analysis/analyzers/js/analyzer.ts [hunk 2]
```diff
@@ -51,10 +51,12 @@ import { JsTsAnalysisInput, JsTsAnalysisOutput } from './analysis';
...
+  const linter = getLinter(input.linterId || 'default');
```
Comment (vilchik-elena) [inline eslint-bridge/src/services/analysis/analyzers/js/analyzer.ts:55]: we don't need `|| 'default'` as this parameter has a default value

File: eslint-bridge/src/services/analysis/analyzers/yaml/analyzer.ts [hunk 3]
```diff
@@ -52,7 +53,8 @@ export const EMPTY_YAML_ANALYSIS_OUTPUT: YamlAnalysisOutput = {
  * @returns the YAML analysis output
  */
 export function analyzeYAML(input: YamlAnalysisInput): YamlAnalysisOutput {
-  assertLinterInitialized();
+  debug(`analyzing file linterId=${input.linterId}`);
```
Comment (vilchik-elena) [inline eslint-bridge/src/services/analysis/analyzers/yaml/analyzer.ts:56]: same here, log file path

File: eslint-bridge/src/services/analysis/analyzers/yaml/analyzer.ts [hunk 4]
```diff
@@ -52,7 +53,8 @@ export const EMPTY_YAML_ANALYSIS_OUTPUT: YamlAnalysisOutput = {
...
+  const linter = getLinter(input.linterId || 'default');
```
Comment (vilchik-elena) [inline eslint-bridge/src/services/analysis/analyzers/yaml/analyzer.ts:57]: same here, no need for default

File: eslint-bridge/src/linting/eslint/index.ts [hunk 5]
```diff
@@ -81,3 +75,17 @@ function loadBundles(bundles: string[]) {
...
+/**
+ * Returns the linter with the given ID
+ *
+ * @param linterId key of the linter
+ *
+ * Throws a runtime error if the global linter wrapper is not initialized.
+ */
+export function getLinter(linterId: keyof Linters = 'default') {
+  if (!linters[linterId]) {
+    throw APIError.linterError(`Linter ${linterId} does not exist. Did you call /init-linter?`);
```
Comment (vilchik-elena) [inline eslint-bridge/src/linting/eslint/index.ts:88]: let's add quotes around `${linterId}`

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java [hunk 6]
```diff
@@ -22,39 +22,66 @@
...
   @TempDir
   private Path gitBaseDir;
 
+  @BeforeAll
+  public static void startOrchestrator() {
+    orchestrator = OrchestratorStarter.getOrchestratorBuilder()
+      .setEdition(Edition.DEVELOPER).activateLicense()
+      .addPlugin(MavenLocation.of("com.sonarsource.security", "sonar-security-plugin", "LATEST_RELEASE"))
+      .addPlugin(MavenLocation.of("com.sonarsource.security", "sonar-security-js-frontend-plugin", "LATEST_RELEASE"))
+      .addPlugin(MavenLocation.of("com.sonarsource.security", "sonar-security-php-frontend-plugin", "LATEST_RELEASE"))
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java:70]: don't use `getOrchestratorBuilder`, then you can drop php frontend

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java [hunk 7]
```diff
@@ -109,49 +136,96 @@ private static BuildResult scanWith(SonarScanner scanner) {
...
 
     try (var gitExecutor = createProjectIn(projectPath, language)) {
       gitExecutor.execute(git -> git.checkout().setName(Master.BRANCH));
-      assertThat(scanWith(getMasterScannerIn(projectPath, projectKey))).has(allOf(
-        expectedLog("DEBUG: Saving issue for rule no-extra-semi", Master.ANALYZER_REPORTED_ISSUES),
-        expectedLog(String.format("INFO: %1$d/%1$d source files have been analyzed", Master.SOURCE_FILES), 1)
-      ));
-      assertThat(getIssues(projectKey))
+      BuildResultAssert.assertThat(scanWith(getMasterScannerIn(projectPath, projectKey)))
+        .hasLogs("DEBUG: Saving issue for rule no-extra-semi", Master.ANALYZER_REPORTED_ISSUES)
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java:162]: let's order logs in the order of appearence

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java [hunk 8]
```diff
@@ -109,49 +136,96 @@ private static BuildResult scanWith(SonarScanner scanner) {
...
+        .hasAtLeastLog("Will skip unchanged files")
+        .hasLog("DEBUG: initializing linter \"default\"")
+        .hasLog("DEBUG: initializing linter \"unchanged\"")
+        .hasLog("analyzing file linterId=unchanged")
+        .hasLog("analyzing file linterId=default");
+      assertThat(getIssues(orchestrator, projectKey, PR.BRANCH))
         .hasSize(1)
         .extracting(Issues.Issue::getComponent)
         .contains(projectKey + ":hello." + language);
+      assertThat(findUcfgFilesIn(projectPath)).isNotEmpty();
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java:187]: We should assert that there is at least one UCFG for each file

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java [hunk 9]
```diff
@@ -109,49 +136,96 @@ private static BuildResult scanWith(SonarScanner scanner) {
...
+
+    BuildResultAssert hasNotLog(String log) {
+      return hasLogs(log, 0);
+    }
+
+    BuildResultAssert hasLogs(String log, int times) {
+      return has(new HamcrestCondition<>(logMatcher(String.format("has logs [%s] %d time(s)", log, times), log, n -> n == times)));
+    }
+
+    BuildResultAssert hasAtLeastLog(String log) {
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java:227]: `logsAtLeastOnce`

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java [hunk 10]
```diff
@@ -109,49 +136,96 @@ private static BuildResult scanWith(SonarScanner scanner) {
...
+
+    BuildResultAssert hasLog(String log) {
+      return hasLogs(log, 1);
+    }
+
+    BuildResultAssert hasNotLog(String log) {
+      return hasLogs(log, 0);
+    }
+
+    BuildResultAssert hasLogs(String log, int times) {
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java:223]: `logsTimes`

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java [hunk 11]
```diff
@@ -109,49 +136,96 @@ private static BuildResult scanWith(SonarScanner scanner) {
...
+            .isPresent();
+        }
+      };
+    }
+
+    BuildResultAssert hasLog(String log) {
+      return hasLogs(log, 1);
+    }
+
+    BuildResultAssert hasNotLog(String log) {
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java:219]: `doesNotLog`

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java [hunk 12]
```diff
@@ -109,49 +136,96 @@ private static BuildResult scanWith(SonarScanner scanner) {
...
+            .filter(BuildResult.class::isInstance)
+            .map(BuildResult.class::cast)
+            .map(result -> result.getLogsLines(line -> line.contains(log)).size())
+            .filter(predicate)
+            .isPresent();
+        }
+      };
+    }
+
+    BuildResultAssert hasLog(String log) {
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java:215]: `logsOnce`

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java [hunk 13]
```diff
@@ -109,49 +136,96 @@ private static BuildResult scanWith(SonarScanner scanner) {
...
+        .hasLog("analyzing file linterId=default");
+      assertThat(getIssues(orchestrator, projectKey, PR.BRANCH))
         .hasSize(1)
         .extracting(Issues.Issue::getComponent)
         .contains(projectKey + ":hello." + language);
+      assertThat(findUcfgFilesIn(projectPath)).isNotEmpty();
+    }
+  }
+
+  static class BuildResultAssert extends AbstractAssert<BuildResultAssert, BuildResult> {
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java:191]: I suggest to extract this class into another file. This piece of logic is pretty big and independent, we could reuse it later

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java [hunk 14]
```diff
@@ -109,49 +136,96 @@ private static BuildResult scanWith(SonarScanner scanner) {
...
-        expectedLog(String.format("INFO: %1$d/%1$d source files have been analyzed", Master.SOURCE_FILES), 1)
-      ));
-      assertThat(getIssues(projectKey))
+      BuildResultAssert.assertThat(scanWith(getMasterScannerIn(projectPath, projectKey)))
+        .hasLogs("DEBUG: Saving issue for rule no-extra-semi", Master.ANALYZER_REPORTED_ISSUES)
+        .hasLog(String.format("INFO: %1$d/%1$d source files have been analyzed", Master.SOURCE_FILES))
+        .hasAtLeastLog("INFO: Won't skip unchanged files as this is not activated in the sensor context")
+        .hasLog("DEBUG: initializing linter \"default\"")
+        .hasNotLog("DEBUG: initializing linter \"unchanged\"")
+        .hasLogs("analyzing file linterId=default", Master.SOURCE_FILES);
```
Comment (vilchik-elena) [inline its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java:167]: let's add log level to add logs here

File: eslint-bridge/tests/services/analysis/analyzers/js/analyzer.test.ts [hunk 15]
```diff
@@ -43,52 +43,68 @@ describe('analyzeJSTS', () => {
     );
   });
 
-  it('should analyze JavaScript code', () => {
+  it('should analyze JavaScript code with the given linter', () => {
     const rules = [
       { key: 'prefer-default-last', configurations: [], fileTypeTarget: ['MAIN'] },
     ] as RuleConfig[];
     initializeLinter(rules);
+    initializeLinter([], [], [], 'empty');
```
Comment (vilchik-elena) [inline eslint-bridge/tests/services/analysis/analyzers/js/analyzer.test.ts:51]: I don't understand why you decided to add this test case

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+    if (!isRuntimeApiCompatible(context)) {
+      LOG.info("Won't skip unchanged files as the API is not compatible");
+      skipUnchangedFiles = false;
+      unchangedFileRules = List.of();
+      return;
+    }
+
+    var canSkipUnchangedFiles = context.canSkipUnchangedFiles();
+    if (!canSkipUnchangedFiles) {
+      LOG.info("Won't skip unchanged files as this is not activated in the sensor context");
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java:71]: this message should be in Debug and use more user-friendly terms

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+      skipUnchangedFiles = false;
+      unchangedFileRules = List.of();
+      return;
+    }
+
+    var containsUcfgRule = EslintRule.containsRuleWithKey(rules, EslintRule.UCFG_ESLINT_KEY);
+    if (!containsUcfgRule) {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java:78]: add comment that this theoretically should never happen as PR analysis exists only on Developer edition where UCFG rule should be available

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+      LOG.info("Won't skip unchanged files as there's no rule with the ESLint key '{}'", EslintRule.UCFG_ESLINT_KEY);
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java:79]: in fact it's opposite, we will fully skip unchanged files in this case

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    var canSkipUnchangedFiles = context.canSkipUnchangedFiles();
+    if (!canSkipUnchangedFiles) {
+      LOG.info("Won't skip unchanged files as this is not activated in the sensor context");
+      skipUnchangedFiles = false;
+      unchangedFileRules = List.of();
+      return;
+    }
+
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java:76]: I suggest to add here an INFO log `Performing optimized analysis as we can skip files which didn't change`

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java (hunk 18 above)
Comment (vilchik-elena) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java:79]: I suggest  `LOG.info("Files which didn't change are excluded from analysis")`

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+    var containsUcfgRule = EslintRule.containsRuleWithKey(rules, EslintRule.UCFG_ESLINT_KEY);
+    if (!containsUcfgRule) {
+      LOG.info("Won't skip unchanged files as there's no rule with the ESLint key '{}'", EslintRule.UCFG_ESLINT_KEY);
+      skipUnchangedFiles = true;
+      unchangedFileRules = List.of();
+      return;
+    }
+
+    LOG.info("Will skip unchanged files");
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java:85]: I suggest  `LOG.info("Files which didn't change will be part of UCFG generation only, other rules will not be executed")`

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+    initializeIfNeeded();
+    return skipUnchangedFiles && !unchangedFileRules.isEmpty();
+  }
+
+  List<EslintRule> getUnchangedFileRules() {
+    initializeIfNeeded();
+    return unchangedFileRules;
+  }
+
+  List<InputFile> getFilesToAnalyzeIn(List<InputFile> inputFiles) {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java:100]: I am not sure why name has `in` at the end. I suggest to drop it

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+    initializeIfNeeded();
+    // IF we can skip unchanged files AND there's no rule for unchanged files THEN we can analyse only changed files.
+    if (skipUnchangedFiles && unchangedFileRules.isEmpty()) {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java:103]: let's use enum of the analysis mode to not compute it here again

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/EslintRule.java [hunk 23]
```diff
@@ -34,6 +38,22 @@ class EslintRule {
     this.configurations = configurations;
   }
 
+  static boolean containsRuleWithKey(List<EslintRule> rules, String eslintKey) {
+    return rules.stream().anyMatch(ruleMatcher(eslintKey));
+  }
+
+  static List<EslintRule> findFirstRuleWithKey(List<EslintRule> rules, String eslintKey) {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/EslintRule.java:45]: let's just return a rule, not the list

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/YamlSensor.java [hunk 24]
```diff
@@ -61,12 +63,13 @@ public void describe(SensorDescriptor descriptor) {
   }
 
   @Override
-  protected void analyzeFiles(List<InputFile> inputFiles) throws IOException {
+  protected void analyzeFiles(List<InputFile> allFiles) throws IOException {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/YamlSensor.java:66]: let's add yaml to the integration test

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java (hunk 6 above)
Comment (francoismora) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java]: Done.

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java (hunk 7 above)
Comment (francoismora) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java]: Done

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java (hunk 8 above)
Comment (francoismora) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java]: Done

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java (hunk 9 above)
Comment (francoismora) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java]: Done

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java (hunk 10 above)
Comment (francoismora) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java]: Done

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java (hunk 11 above)
Comment (francoismora) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java]: Done

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java (hunk 12 above)
Comment (francoismora) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java]: Done

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java (hunk 13 above)
Comment (francoismora) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java]: Done

File: its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java (hunk 14 above)
Comment (francoismora) [reply its/plugin/tests/src/test/java/com/sonar/javascript/it/plugin/PRAnalysisTest.java]: Done

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+class AnalysisOptions {
+
+  static final String UNCHANGED_LINTER_ID = "unchanged";
+  static final String DEFAULT_LINTER_ID = "default";
+  private static final Logger LOG = Loggers.get(AnalysisOptions.class);
+  private final Initializer initializer;
+  private boolean initialized = false;
+  private boolean skipUnchangedFiles;
+  private List<EslintRule> unchangedFileRules;
+  AnalysisOptions(SensorContext context, List<EslintRule> rules) {
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java]: Done

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java (hunk 16 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java]: Done

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java (hunk 17 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java]: Done

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java (hunk 18 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java]: Done

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java (hunk 19 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java]: As discussed this is not needed anymore.

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java (hunk 20 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java]: Done

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java (hunk 21 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java]: As discussed this is not needed anymore.

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java (hunk 22 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisOptions.java]: Done

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/EslintRule.java (hunk 23 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/EslintRule.java]: Done

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/AnalysisModeTest.java [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+class AnalysisModeTest {
+
+  @Mock
+  SensorContext context;
+  private AutoCloseable mocks;
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/AnalysisModeTest.java:50]: I don't understand why we need this one

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/AnalysisModeTest.java [hunk 27]
```diff
@@ -0,0 +1,190 @@
...
+    when(context.canSkipUnchangedFiles()).thenReturn(false);
+
+    var rules = rules("key1", "key2", "ucfg");
+    var mode = AnalysisMode.getModeFor(context, rules);
+    assertThat(mode).isEqualTo(AnalysisMode.DEFAULT);
+    assertThat(mode.getUnchangedFileRules(rules)).isEqualTo(rules);
+    verify(context).canSkipUnchangedFiles();
+
+    var files = fileList(changedFiles(2), unchangedFiles(1), addedFiles(3));
+    for (var file : files) {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/AnalysisModeTest.java:112]: can't you do same as on line 97?

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBridgeServerImplTest.java [hunk 28]
```diff
@@ -174,9 +175,9 @@ void test_init() throws Exception {
     eslintBridgeServer.startServer(context, emptyList());
 
     List<EslintRule> rules = Collections.singletonList(new EslintRule("key", singletonList("config"), Collections.singletonList(InputFile.Type.MAIN)));
-    eslintBridgeServer.initLinter(rules, Collections.emptyList(), Collections.emptyList());
+    eslintBridgeServer.initLinter(rules, Collections.emptyList(), Collections.emptyList(), AnalysisMode.getModeFor(context, Collections.emptyList()));
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBridgeServerImplTest.java:178]: I don't see the point to call for `getModeFor` here, can we hardcode the value?

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisMode.java [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+import static java.util.Collections.emptyList;
+
+enum AnalysisMode {
+  DEFAULT, SKIP_UNCHANGED;
+
+  static final String UNCHANGED_LINTER_ID = "unchanged";
+  static final String DEFAULT_LINTER_ID = "default";
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisMode.java:35]: Done

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisMode.java [hunk 30]
```diff
@@ -0,0 +1,86 @@
...
+    if (this == SKIP_UNCHANGED) {
+      var rule = EslintRule.findFirstRuleWithKey(rules, EslintRule.UCFG_ESLINT_KEY);
+      return rule == null ? emptyList() : List.of(rule);
+    } else {
+      return rules;
+    }
+  }
+
+  String getLinterIdFor(InputFile file) {
+    // IF we can skip unchanged files AND the file is unchanged THEN we can use the unchanged linter.
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisMode.java:78]: Removed

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisMode.java [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+      LOG.debug(logDefaultMode, "security rules are not available");
+      return AnalysisMode.DEFAULT;
+    }
+
+    LOG.debug("Files which didn't change will be part of UCFG generation only, other rules will not be executed");
+    return AnalysisMode.SKIP_UNCHANGED;
+  }
+
+  List<EslintRule> getUnchangedFileRules(List<EslintRule> rules) {
+    if (this == SKIP_UNCHANGED) {
```
[the remark this answers is not in the document]
Comment (francoismora) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/AnalysisMode.java:69]: Dropped

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/AnalysisModeTest.java (hunk 26 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/AnalysisModeTest.java:50]: That's taken from Mockito annotations
image

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/AnalysisModeTest.java (hunk 27 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/AnalysisModeTest.java:112]: Done

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBridgeServerImplTest.java (hunk 28 above)
Comment (francoismora) [reply sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/eslint/EslintBridgeServerImplTest.java:178]: Hard-coded

**Review by vilchik-elena — APPROVED** [review]

**Pull request merged**